optuna/samplers/_tpe/parzen_estimator.py

```diff
@@ -18,114 +33,408 @@ class _ParzenEstimatorParameters(
+        return samples_dict
+
+    def log_pdf(self, samples_dict: Dict[str, np.ndarray]) -> np.ndarray:
+
+        samples_dict = self._transform_to_uniform(samples_dict)
+        n_observations = len(self._weights)
+        n_samples = next(iter(samples_dict.values())).size
+        if n_samples == 0:
+            return np.asarray([], dtype=float)
+        # We compute log pdf (component_log_pdf)
+        # for each sample in samples_dict (of size n_samples)
+        # for each component of `_MultivariateParzenEstimator` (of size n_observations).
+        component_log_pdf = np.zeros((n_samples, n_observations))
+        for param_name, dist in self._search_space.items():
+            samples = samples_dict[param_name]
+            if isinstance(dist, distributions.CategoricalDistribution):
+                categorical_weights = self._categorical_weights[param_name]
+                assert categorical_weights is not None
+                log_pdf = np.log(categorical_weights.T[samples, :])
+            else:
+                # We restore parameters of parzen estimators.
+                low = self._low[param_name]
+                high = self._high[param_name]
+                q = self._q[param_name]
+                mus = self._mus[param_name]
+                sigmas = self._sigmas[param_name]
+                assert low is not None
+                assert high is not None
+                assert mus is not None
+                assert sigmas is not None
+
+                cdf_func = _ParzenEstimator._normal_cdf
+                p_accept = cdf_func(high, mus, sigmas) - cdf_func(low, mus, sigmas)
+                if q is None:
+                    distance = samples[:, None] - mus
+                    mahalanobis = distance / np.maximum(sigmas, EPS)
+                    z = np.sqrt(2 * np.pi) * sigmas
+                    coefficient = 1 / z / p_accept
+                    log_pdf = -0.5 * mahalanobis ** 2 + np.log(coefficient)
+                else:
+                    upper_bound = np.minimum(samples + q / 2.0, high)
+                    lower_bound = np.maximum(samples - q / 2.0, low)
+                    cdf = cdf_func(upper_bound[:, None], mus[None], sigmas[None]) - cdf_func(
+                        lower_bound[:, None], mus[None], sigmas[None]
+                    )
+                    log_pdf = np.log(cdf + EPS) - np.log(p_accept + EPS)
+            component_log_pdf += log_pdf
+        ret = scipy.special.logsumexp(component_log_pdf + np.log(self._weights), axis=1)
+        return ret
+
+    def _calculate_weights(self) -> np.ndarray:
+
+        # We decide the weights.
+        consider_prior = self._parameters.consider_prior
+        prior_weight = self._parameters.prior_weight
+        weights_func = self._parameters.weights
+        n_observations = self._n_observations
+        if consider_prior:
+            weights = np.zeros(n_observations + 1)
+            weights[:-1] = weights_func(n_observations)[:n_observations]
```

Comment:
notes: Different implementation from `multivariate_parzen_estimator`.

Reply:
I think we can leave a TODO to raise if the weight function is returning an ndarray of unexpected size.